simd/src/Kokkos_SIMD_NEON.hpp

```diff
@@ -0,0 +1,1108 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SIMD_NEON_HPP
+#define KOKKOS_SIMD_NEON_HPP
+
+#include <functional>
+#include <type_traits>
+
+#include <Kokkos_SIMD_Common.hpp>
+
+#include <arm_neon.h>
+
+namespace Kokkos {
+
+namespace Experimental {
+
+namespace simd_abi {
+
+template <int N>
+class neon_fixed_size {};
+
+}  // namespace simd_abi
+
+template <>
+class simd_mask<double, simd_abi::neon_fixed_size<2>> {
+  uint64x2_t m_value;
+
+ public:
+  class reference {
+    uint64x2_t& m_mask;
+    int m_lane;
+
+   public:
+    KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION reference(uint64x2_t& mask_arg,
+                                                    int lane_arg)
+        : m_mask(mask_arg), m_lane(lane_arg) {}
+    KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION reference
+    operator=(bool value) const {
+      // this switch statement is needed because the lane argument has to be a
+      // constant
+      switch (m_lane) {
+        case 0:
+          m_mask = vsetq_lane_u64(value ? 0xFFFFFFFFFFFFFFFFULL : 0, m_mask, 0);
+          break;
+        case 1:
+          m_mask = vsetq_lane_u64(value ? 0xFFFFFFFFFFFFFFFFULL : 0, m_mask, 1);
+          break;
+      }
```

Comment:
You cannot simply do
```c++
m_mask = vsetq_lane_u64(value ? 0xFFFFFFFFFFFFFFFFULL : 0, m_mask, (int const) m_lane);
```
or
```c++
int const const_lane = m_lane ? 0 : 1;
m_mask = vsetq_lane_u64(value ? 0xFFFFFFFFFFFFFFFFULL : 0, m_mask, const_lane);
```
It is easier to see what's the difference between the two calls to `vsetq_lane_u64`

Reply:
Yea, that's what I had initially, and the ARM clang compiler gave an error saying that the lane argument must be a compile-time constant.... yea I fully agree with you that is bad design, but the compiler forces us.